crates/client/l1-messages/src/worker.rs

```diff
@@ -0,0 +1,192 @@
+use std::sync::Arc;
+
+use ethers::core::types::Address;
+use ethers::providers::{Http, Provider, StreamExt};
+use madara_runtime::pallet_starknet;
+use mp_transactions::HandleL1MessageTransaction;
+use pallet_starknet::runtime_api::{ConvertTransactionRuntimeApi, StarknetRuntimeApi};
+use parity_scale_codec::{Decode, Encode};
+use sc_client_api::{Backend, HeaderBackend};
+use sc_transaction_pool_api::{TransactionPool, TransactionSource};
+use sp_api::offchain::OffchainStorage;
+use sp_api::{BlockId, ProvideRuntimeApi};
+use sp_runtime::traits::Block as BlockT;
+use starknet_api::transaction::Fee;
+
+use crate::config::L1MessagesWorkerConfig;
+use crate::contract::{L1Contract, LogMessageToL2Filter};
+use crate::error::L1MessagesWorkerError;
+
+const TX_SOURCE: TransactionSource = TransactionSource::External;
+const STORAGE_PREFIX: &str = "L1MessagesWorker";
+const STORAGE_KEY: &str = "last_synced_block";
+
+pub fn connect_to_l1_contract(
+    config: &L1MessagesWorkerConfig,
+) -> Result<L1Contract<Provider<Http>>, L1MessagesWorkerError> {
+    let address: Address =
+        config.contract_address.parse::<Address>().map_err(|_| L1MessagesWorkerError::ConfigError)?;
+    let provider = Provider::<Http>::try_from(&config.http_provider).map_err(|e| {
+        log::error!("⟠ Failed to connect to L1 Node: {:?}", e);
+        L1MessagesWorkerError::ConfigError
+    })?;
+
+    let l1_contract = L1Contract::new(address, Arc::new(provider));
+
+    Ok(l1_contract)
+}
+
+pub async fn run_worker<C, P, B, S>(config: L1MessagesWorkerConfig, client: Arc<C>, pool: Arc<P>, backend: Arc<S>)
+where
+    B: BlockT,
+    C: ProvideRuntimeApi<B> + HeaderBackend<B>,
+    C::Api: StarknetRuntimeApi<B> + ConvertTransactionRuntimeApi<B>,
+    P: TransactionPool<Block = B> + 'static,
+    S: Backend<B> + Send + 'static,
+{
+    log::info!("⟠ Starting L1 Messages Worker with settings: {:?}", config);
+
+    let mut offchain_storage = match backend.offchain_storage() {
+        Some(offchain_storage) => offchain_storage,
+        None => {
+            log::error!("⟠ Offchain storage unavailable");
+            return;
+        }
+    };
+
+    let l1_contract = match connect_to_l1_contract(&config) {
+        Ok(l1_contract) => l1_contract,
+        Err(e) => {
+            log::error!("⟠ Unexpected error while connecting to L1: {:?}", e);
+            return;
+        }
+    };
+
+    let events = l1_contract.events().from_block(get_block_number::<B, S>(&offchain_storage).unwrap_or_default());
```

Comment:
here `events` is a finite iterator with a finite length. You make it a stream, so once we went through this finite number of events, it will just yield `None` forever.

That is not the behavior we want. We want it to keep looking for new l1 blocks and process their events

Reply:
@tdelabro but it keeps looking for new blocks...